dashboard/app/helpers/users_helper.rb

```diff
@@ -1,4 +1,106 @@
 module UsersHelper
+  include ApplicationHelper
+
+  # Returns the number of lines written in the current user session.
+  # return [Integer]
+  def session_lines
+    session[:lines] || 0
+  end
+
+  # Add additional lines completed in the given session
+  # @param [Integer] added_lines
+  def session_add_lines(added_lines)
+    session[:lines] = session_lines + added_lines
+  end
+
+  # Resets all user session state (level progress, lines of code, videos seen,
+  # etc.) for tests.
+  def session_reset_for_test
+    session[:lines] = nil
+    session[:progress] = nil
+    session[:callouts_seen] = nil
+    session[:videos_seen] = nil
+  end
+
+  # Returns the progress value for the given level_id, or 0 if there
+  # has been no progress.
+  # @param [Integer] level_id
+  # return [Integer]
+  def session_level_progress(level_id)
+    (session[:progress] || {}).fetch(level_id.to_i, 0)
+  end
+
+  # Sets the progress for the given level_id for the current session.
+  # @param [Integer] level_id
+  # @param [Integer] progress
+  def session_set_level_progress(level_id, progress)
+    session[:progress] ||= {}
+    session[:progress][level_id.to_i] = progress
+  end
+
+  # Returns true if there has been no progress in completing levels for
+  # the current session.
+  def session_levels_progress_is_empty_for_test
+    !session[:progress] || session[:progress].empty?
+  end
+
+  # Adds 'script' to the set of scripts completed for the current session.
+  # @param [Integer] script_id
+  def session_add_script(script_id)
+    scripts = session_scripts
+    scripts << script_id.to_i unless scripts.include?(script_id)
+    session[:scripts] = scripts
+  end
+
+  # Returns an array of ids of the scripts completed in the current session.
+  # Callers should not mutate the array.
+  # @return [Array<Integer>]
+  def session_scripts
```

Comment:
add_script is used in activities_controllers.

The script accessor is only used in tests at the moment.